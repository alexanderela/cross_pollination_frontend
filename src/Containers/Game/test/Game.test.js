import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import {Game, mapStateToProps, mapDispatchToProps} from '../index';
// import { wrap } from 'module';

describe('Game', () => {
  let wrapper;
  let mockCorrectCountry;
  let addPoints;
  let mockEvent;
  let mockUser;

  beforeEach(() => {
    mockCorrectCountry = {
                            "id": 46,
                            "name": "The Netherlands",
                            "flag": "/images/flags/netherlands.png",
                            "country_outline": "/images/outlines/netherlands.png",
                            "created_at": "2018-12-28T12:02:21.458Z",
                            "updated_at": "2018-12-28T12:02:21.458Z",
                            "multipleChoice": [
                              "Seychelles",
                              "Malawi",
                              "Cambodia",
                              "The Netherlands"
                            ],
                            "facts": {
                              "id": 106,
                              "country_fact": "Numerous dikes cover the coast of Ijsselmeer in this country",
                              "country_id": 46,
                              "created_at": "2018-12-28T12:02:21.951Z",
                              "updated_at": "2018-12-28T12:02:21.951Z"
                            }
                          };

    mockEvent = { target: { innerText: 'The Netherlands' } };

    mockUser = { id: 1, name: 'Ben', loggedIn: true}

    wrapper = shallow(<Game 
                        compilePoints={jest.fn()}  
                        totalPoints={10}
                        getCountry={jest.fn()}
                        currentCountry={mockCorrectCountry}
                        user={mockUser}
                      />);
    
  });

  it('should render like the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('checkAnswer', () => {
    it('should set state if the answer is correct', () => {
      wrapper.setState({ status: '' });
      wrapper.instance().addPoints = jest.fn()

      wrapper.instance().checkAnswer(mockEvent);
      expect(wrapper.state().status).toEqual('Correct');
    });

    it('should invoke addPoints if answer is correct', () => {
      wrapper.instance().addPoints = jest.fn()

      wrapper.instance().checkAnswer(mockEvent);
      expect(wrapper.instance().addPoints).toHaveBeenCalled
    })

    it('should set state if the answer is incorrect', () => {
      wrapper.setState({ status: '' });
      mockEvent = { target: { innerText: 'France' } };

      wrapper.instance().checkAnswer(mockEvent);
      expect(wrapper.state().status).toEqual('Wrong');
    });
  });

  describe('giveHint', () => {
    it('should set state to show a hint', () => {
      wrapper.setState({ showHint: false });
      wrapper.instance().giveHint()
      expect(wrapper.state().showHint).toEqual(true)     
    })

    it('should set state if user requests a hint', () => {
      wrapper.setState({ hint: '', hintsUsed: 0 });
      wrapper.instance().giveHint()
      expect(wrapper.state().hint).toEqual('fact')
    });

    it('should set state if user requests a second hint', () => {
      wrapper.setState({ hint: '', hintsUsed: 1 });
      wrapper.instance().giveHint()
      expect(wrapper.state().hint).toEqual('outline')
    });

    it('should exhaust all hints after 2 hints are given', () => {
      wrapper.setState({ hint: '', hintsUsed: 2 });
      wrapper.instance().giveHint()
      expect(wrapper.state().hint).toEqual('out of hints')
      expect(wrapper.state().hintsExhausted).toEqual(true)
    });

    it('should set new state with updated hintsUsed and pointsPossible counts', () => {
      wrapper.setState({ 
                         hint: '', 
                         hintsUsed: 0, 
                         pointsPossible: 3 
                       });

      wrapper.instance().giveHint()
      expect(wrapper.state().hintsUsed).toEqual(1)
      expect(wrapper.state().pointsPossible).toEqual(2)
    })
  });

  describe('showButtons', () => {
    it('should show buttons if there are country choice options', () => {
      wrapper.instance().showButtons()

      expect(wrapper.find('.option-button')).toHaveLength(4);
    })
    it('should not show buttons if there are no country choice options', () => {
      mockCorrectCountry = {
                        "id": 46,
                        "name": "The Netherlands",
                        "flag": "/images/flags/netherlands.png",
                        "country_outline": "/images/outlines/netherlands.png",
                        "created_at": "2018-12-28T12:02:21.458Z",
                        "updated_at": "2018-12-28T12:02:21.458Z",
                        "facts": {
                          "id": 106,
                          "country_fact": "Numerous dikes cover the coast of Ijsselmeer in this country",
                          "country_id": 46,
                          "created_at": "2018-12-28T12:02:21.951Z",
                          "updated_at": "2018-12-28T12:02:21.951Z"
                        }
                      };

      mockUser = { id: 1, name: 'Ben', loggedIn: true}
        

        wrapper = shallow(<Game 
                            compilePoints={jest.fn()}  
                            totalPoints={10}
                            getCountry={jest.fn()}
                            currentCountry={mockCorrectCountry}
                            user={mockUser}
                          />);

      wrapper.instance().showButtons()
      expect(wrapper.find('.option-button')).toHaveLength(0);
    })
  })

  describe('closeResults', () => {
    it('should close results and reset for next country', () => {
      wrapper.setState({
        hintsExhausted: true,
        hintsUsed: 1,
        pointsPossible: 2,
        status: 'Wrong'
      })

      wrapper.instance().closeResults()
      expect(wrapper.state().hintsExhausted).toEqual(false)
      expect(wrapper.state().hintsUsed).toEqual(0)
      expect(wrapper.state().pointsPossible).toEqual(3)
      expect(wrapper.state().status).toEqual('')
    })
  })  

  describe('addPoints', () => {
    it('should set state upon correct guess', () => {
      wrapper.setState({
        pointsPossible: 3,
        totalPoints: 0,
        status: 'Correct'
      })
      wrapper.instance().addPoints();
      expect(wrapper.state().totalPoints).toEqual(3)
    })

    it('should set state upon incorrect guess', () => {
      wrapper.setState({
        pointsPossible: 3,
        totalPoints: 0,
        status: 'Wrong'
      })
      wrapper.instance().addPoints();
      expect(wrapper.state().totalPoints).toEqual(0)
    })
  })

  describe('hideHint', () => {
    it('should hide each hint', () => {
      wrapper.setState({ showHint: true })

      wrapper.instance().hideHint()
      expect(wrapper.state().showHint).toEqual(false)   
    })
  })
  
 describe('getCountryFlagPath', () => {
    it('should create complete path for country flag', () => {
      let result = wrapper.instance().getCountryFlagPath()
      expect(result).toEqual('https://flagz4u.herokuapp.com/images/flags/netherlands.png')
    })
  })

  describe('mapStateToProps', () => {
    let mockState = {
      user: {user: 'Bob', email: "bob@bob.com"},
      currentCountry: {
        "id": 79,
        "name": "Turkmenistan",
        "flag": "/images/flags/turkmenistan.png",
        "country_outline": "/images/outlines/turkmenistan.png",
        "multipleChoice": [
          "France",
          "Nigeria",
          "Turkmenistan",
          "Japan"
        ]
      },
      usedCountries: ['Mexico', 'Hungary', 'Ireland', 'Sweden'],
    }

    it("should return a currentCountry in the props object", () => {
      const expected = {
        "id": 79,
        "name": "Turkmenistan",
        "flag": "/images/flags/turkmenistan.png",
        "country_outline": "/images/outlines/turkmenistan.png",
        "multipleChoice": [
          "France",
          "Nigeria",
          "Turkmenistan",
          "Japan"
        ]
      }
        
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps.currentCountry).toEqual(expected)
    })
  });
});

