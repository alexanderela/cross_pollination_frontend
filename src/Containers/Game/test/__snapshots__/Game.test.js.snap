// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render like the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Game
    compilePoints={[MockFunction]}
    currentCountry={
      Object {
        "country_outline": "/images/outlines/netherlands.png",
        "created_at": "2018-12-28T12:02:21.458Z",
        "facts": Object {
          "country_fact": "Numerous dikes cover the coast of Ijsselmeer in this country",
          "country_id": 46,
          "created_at": "2018-12-28T12:02:21.951Z",
          "id": 106,
          "updated_at": "2018-12-28T12:02:21.951Z",
        },
        "flag": "/images/flags/netherlands.png",
        "id": 46,
        "multipleChoice": Array [
          "Seychelles",
          "Malawi",
          "Cambodia",
          "The Netherlands",
        ],
        "name": "The Netherlands",
        "updated_at": "2018-12-28T12:02:21.458Z",
      }
    }
    getCountry={[MockFunction]}
    totalPoints={10}
    user={
      Object {
        "id": 1,
        "loggedIn": true,
        "name": "Ben",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="top-container"
        >
          <div
            className="account-area"
          >
            <div
              className="back-button"
            />
            <h5
              className="account-text"
            >
              Ben
               
              <strong>
                0
              </strong>
            </h5>
          </div>
        </div>,
        <div
          className="flag-main"
        >
          <img
            className="flag-image"
            src="https://flagz4u.herokuapp.com/images/flags/netherlands.png"
          />
        </div>,
        <div
          className="hint-button"
          onClick={[Function]}
        >
          Hints: 2
        </div>,
        Array [
          <div
            className="option-button button"
            name="Seychelles"
            onClick={[Function]}
          >
            Seychelles
          </div>,
          <div
            className="option-button button"
            name="Malawi"
            onClick={[Function]}
          >
            Malawi
          </div>,
          <div
            className="option-button button"
            name="Cambodia"
            onClick={[Function]}
          >
            Cambodia
          </div>,
          <div
            className="option-button button"
            name="The Netherlands"
            onClick={[Function]}
          >
            The Netherlands
          </div>,
        ],
        false,
        false,
      ],
      "className": "Game",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="account-area"
          >
            <div
              className="back-button"
            />
            <h5
              className="account-text"
            >
              Ben
               
              <strong>
                0
              </strong>
            </h5>
          </div>,
          "className": "top-container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="back-button"
              />,
              <h5
                className="account-text"
              >
                Ben
                 
                <strong>
                  0
                </strong>
              </h5>,
            ],
            "className": "account-area",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "back-button",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Ben",
                  " ",
                  <strong>
                    0
                  </strong>,
                ],
                "className": "account-text",
              },
              "ref": null,
              "rendered": Array [
                "Ben",
                " ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": 0,
                  },
                  "ref": null,
                  "rendered": 0,
                  "type": "strong",
                },
              ],
              "type": "h5",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <img
            className="flag-image"
            src="https://flagz4u.herokuapp.com/images/flags/netherlands.png"
          />,
          "className": "flag-main",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "flag-image",
            "src": "https://flagz4u.herokuapp.com/images/flags/netherlands.png",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Hints: 2",
          "className": "hint-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Hints: 2",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "Seychelles",
        "nodeType": "host",
        "props": Object {
          "children": "Seychelles",
          "className": "option-button button",
          "name": "Seychelles",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Seychelles",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "Malawi",
        "nodeType": "host",
        "props": Object {
          "children": "Malawi",
          "className": "option-button button",
          "name": "Malawi",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Malawi",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "Cambodia",
        "nodeType": "host",
        "props": Object {
          "children": "Cambodia",
          "className": "option-button button",
          "name": "Cambodia",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Cambodia",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "The Netherlands",
        "nodeType": "host",
        "props": Object {
          "children": "The Netherlands",
          "className": "option-button button",
          "name": "The Netherlands",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "The Netherlands",
        "type": "div",
      },
      false,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="top-container"
          >
            <div
              className="account-area"
            >
              <div
                className="back-button"
              />
              <h5
                className="account-text"
              >
                Ben
                 
                <strong>
                  0
                </strong>
              </h5>
            </div>
          </div>,
          <div
            className="flag-main"
          >
            <img
              className="flag-image"
              src="https://flagz4u.herokuapp.com/images/flags/netherlands.png"
            />
          </div>,
          <div
            className="hint-button"
            onClick={[Function]}
          >
            Hints: 2
          </div>,
          Array [
            <div
              className="option-button button"
              name="Seychelles"
              onClick={[Function]}
            >
              Seychelles
            </div>,
            <div
              className="option-button button"
              name="Malawi"
              onClick={[Function]}
            >
              Malawi
            </div>,
            <div
              className="option-button button"
              name="Cambodia"
              onClick={[Function]}
            >
              Cambodia
            </div>,
            <div
              className="option-button button"
              name="The Netherlands"
              onClick={[Function]}
            >
              The Netherlands
            </div>,
          ],
          false,
          false,
        ],
        "className": "Game",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="account-area"
            >
              <div
                className="back-button"
              />
              <h5
                className="account-text"
              >
                Ben
                 
                <strong>
                  0
                </strong>
              </h5>
            </div>,
            "className": "top-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="back-button"
                />,
                <h5
                  className="account-text"
                >
                  Ben
                   
                  <strong>
                    0
                  </strong>
                </h5>,
              ],
              "className": "account-area",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "back-button",
                },
                "ref": null,
                "rendered": null,
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Ben",
                    " ",
                    <strong>
                      0
                    </strong>,
                  ],
                  "className": "account-text",
                },
                "ref": null,
                "rendered": Array [
                  "Ben",
                  " ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 0,
                    },
                    "ref": null,
                    "rendered": 0,
                    "type": "strong",
                  },
                ],
                "type": "h5",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              className="flag-image"
              src="https://flagz4u.herokuapp.com/images/flags/netherlands.png"
            />,
            "className": "flag-main",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "flag-image",
              "src": "https://flagz4u.herokuapp.com/images/flags/netherlands.png",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Hints: 2",
            "className": "hint-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Hints: 2",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "Seychelles",
          "nodeType": "host",
          "props": Object {
            "children": "Seychelles",
            "className": "option-button button",
            "name": "Seychelles",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Seychelles",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "Malawi",
          "nodeType": "host",
          "props": Object {
            "children": "Malawi",
            "className": "option-button button",
            "name": "Malawi",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Malawi",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "Cambodia",
          "nodeType": "host",
          "props": Object {
            "children": "Cambodia",
            "className": "option-button button",
            "name": "Cambodia",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Cambodia",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "The Netherlands",
          "nodeType": "host",
          "props": Object {
            "children": "The Netherlands",
            "className": "option-button button",
            "name": "The Netherlands",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "The Netherlands",
          "type": "div",
        },
        false,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
